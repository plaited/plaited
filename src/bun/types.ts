import type {
  BPEvent,
  Actions,
  Trigger,
  BThreads,
  BSync,
  BThread,
  UseSnapshot,
  UseFeedback,
} from '../behavioral.js'
import { Attrs, TemplateObject } from '../jsx/jsx.types.js';

export type ServerWebSocket<T = unknown> = Parameters<
  Extract<Parameters<typeof Bun.serve<T>>[0], { websocket: unknown; unix?: never }>['websocket']['message']
>['0']

export type DefineSocketHandler<T> = (args: {
  send: (address: string) => (event: BPEvent<string>) => ReturnType<ServerWebSocket['send']>
  trigger: Trigger
  useSnapShot: UseSnapshot
  useFeedback: UseFeedback
  bThreads: BThreads
  bSync: BSync
  bThread: BThread
  getContext: () => T
}) => Actions

export type DefineSocketConfig = { publicEvents: string[]; address: string }

export type DefineSocketReturn<T> = {
  address: string
  open(ws: ServerWebSocket<T>): void
  message: (ws: ServerWebSocket<T>, event: BPEvent) => void
}

export type MiddleWareHandler = <T extends Record<string, unknown> = Record<string, unknown>>(
  req: Request,
  ctx?: T & { next: Handler },
) => Promise<Response>

export type Handler = <T extends Record<string, unknown> = Record<string, unknown>>(
  req: Request,
  ctx?: T,
) => Promise<Response>

export type AjaxPageData = [Document, Record<string, Element | string>]


export type Story<T = Attrs> = {
  render: (attrs: T) => TemplateObject
  play: (attrs: T) => void
  description?: string
  a11y: []
  attrs: T
}