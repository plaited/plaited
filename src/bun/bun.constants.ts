export const IGNORED_DIR_PREFIX = '[!_]'
export const GLOB_PATTERN_WORKERS = `**/${IGNORED_DIR_PREFIX}*/worker.@(tsx|ts)`
export const GLOB_PATTERN_MODULES = `**/${IGNORED_DIR_PREFIX}*/module.@(tsx|ts)`
