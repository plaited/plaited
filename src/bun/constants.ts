export const IGNORED_DIR_PREFIX = '[!_]'
export const GLOB_PATTERN_WORKER = `**/${IGNORED_DIR_PREFIX}*/worker.@(tsx|ts)`
export const GLOB_PATTERN_ROUTE = `**/${IGNORED_DIR_PREFIX}*/route.@(tsx|ts)`
export const GLOB_PATTERN_STORIES = `**/*.stories.@(tsx|ts)`