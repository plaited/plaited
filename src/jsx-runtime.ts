/**
 * @internal
 * @module jsx-runtime
 *
 * Purpose: JSX runtime entry point for TypeScript/Babel JSX transformation
 * Architecture: Re-exports createTemplate with standard JSX runtime names
 * Dependencies: createTemplate factory, Fragment component, JSX type definitions
 * Consumers: TypeScript compiler, Babel transforms, bundlers with JSX support

 */

import type { ElementAttributeList } from './main/jsx.types.js'
import { createTemplate, Fragment } from './main/create-template.js'

/**
 * @internal
 * JSX factory exports supporting multiple transform modes:
 * - h: Hyperscript-style factory (classic transform)
 * - jsx: Automatic runtime for single children
 * - jsxs: Automatic runtime for multiple children
 * - jsxDEV: Development mode (same as production currently)
 * All map to same createTemplate for consistency.
 */
export { createTemplate as h, createTemplate as jsx, createTemplate as jsxDEV, createTemplate as jsxs, Fragment }

/**
 * @internal
 * TypeScript JSX namespace declaration.
 * Maps intrinsic elements to Plaited's attribute types.
 * Enables type checking and IntelliSense for JSX elements.
 * IntrinsicElements extends ElementAttributeList which includes
 * all HTML/SVG elements with proper attribute typing.
 */
// eslint-disable-next-line @typescript-eslint/no-namespace
export declare namespace JSX {
  interface IntrinsicElements extends ElementAttributeList {}
}
