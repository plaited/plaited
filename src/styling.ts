export type * from './styling/css.types.js'
export * from './styling/css.js'
export type * from './styling/design-token.types.js'
export * from './styling/get-design-tokens-element.js'
export * from './styling/get-design-tokens-schema.js'
export * from './styling/transform-design-tokens.js'
