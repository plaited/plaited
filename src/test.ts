export * from './test/assert.types.js'
export * from './test/assert.constants.js'
export { assert } from './test/assert.js'
export { findByAttribute } from './test/find-by-attribute.js'
export { findByText } from './test/find-by-text.js'
export { fireEvent } from './test/fire-event.js'
export { match } from './test/match.js'
export { throws } from './test/throws.js'
export { wait } from './utils/wait.js'
export { AssertionError, MissingTestParamsError, TimeoutError } from './test/errors.js'
export { usePlay } from './test/use-play.js'
