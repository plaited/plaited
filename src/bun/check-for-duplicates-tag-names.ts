import type { PlaitedTemplate } from '../client/types.js'
import { isTypeOf } from '../utils/true-type-of.js'
import { PLAITED_TEMPLATE_IDENTIFIER } from '../shared/constants.js'
import { GLOB_PATTERN_DUPLICATE_TAGS } from './bun.constants.js'

const isPlaitedComponent = (mod: unknown): mod is PlaitedTemplate =>
  isTypeOf<PlaitedTemplate>(mod, 'function') && mod?.$ === PLAITED_TEMPLATE_IDENTIFIER

export const checkForDuplicateTagNames = async (dir: string, pattern = GLOB_PATTERN_DUPLICATE_TAGS) => {
  const glob = new Bun.Glob(pattern)
  const components = new Map<string, string[]>()
  const componentPaths = await Array.fromAsync(glob.scan({ cwd: dir }))
  await Promise.all(
    componentPaths.map(async (path) => {
      try {
        const modules = await import(path)
        for (const name of modules) {
          const mod = modules[name]
          if (!mod || !isPlaitedComponent(mod)) continue
          if (components.has(mod.tag)) {
            components.get(mod.tag)?.push(path)
          } else {
            components.set(mod.tag, [path])
          }
        }
        return path
      } catch (err) {
        console.error(err)
      }
    }),
  )
  return [...components].filter(([_, paths]) => paths.length > 1)
}
