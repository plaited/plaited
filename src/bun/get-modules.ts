import { GLOB_PATTERN_MODULES } from './bun.constants.js'
import { isPlaitedTemplate } from './is-plaited-template.js'

const transpiler = new Bun.Transpiler({
  loader: 'tsx',
})

const registry = new Map<string, { filePath: string; exportName: string }>()

const getImports = async (templatePaths: { path: string; kind: string }[]) => {
  const imports = new Set<string>()
  await Promise.all(
    templatePaths.map(async ({ path: filePath }) => {
      try {
        const modules = await import(filePath)
        for (const exportName of modules) {
          const mod = modules[exportName]
          if (!mod || !isPlaitedTemplate(mod)) continue
          if (registry.has(mod.tag)) {
            console.error(
              `Error ${mod.tag} is already registered ${registry.get(mod.tag)?.exportName}@${registry.get(mod.tag)}\n Change tag name of ${exportName}@${filePath}`,
            )
            return
          } else {
            registry.set(mod.tag, { filePath, exportName })
            imports.add(filePath)
          }
        }
      } catch (err) {
        console.error(err)
      }
    }),
  )
  return imports
}

export const getModules = async (dir: string) => {
  const glob = new Bun.Glob(GLOB_PATTERN_MODULES)
  const storyPaths = await Array.fromAsync(glob.scan({ cwd: dir }))
  registry.clear()
  const entries = await Promise.all(
    storyPaths.map(async (filePath) => {
      try {
        const file = Bun.file(filePath)
        const code = await file.text()
        const result = transpiler.scanImports(code)
        const imports = await getImports(result)
        let fileImports = ''
        for (const entry of imports) {
          fileImports += `import '${entry}'\n`
        }
        const regex = /\.tsx?$/
        return { [filePath.replace(regex, '.plaited')]: fileImports }
      } catch (err) {
        console.error(err)
      }
    }),
  )
  return entries
}
