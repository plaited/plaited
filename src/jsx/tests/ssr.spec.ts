/* eslint-disable no-constant-binary-expression */
import { test, expect } from 'bun:test'
import { type FunctionTemplate, type TemplateObject, ssr } from 'plaited'
import { h, Fragment } from 'plaited/jsx-runtime'
import { css } from 'plaited/styling'
import beautify from 'beautify'

const render = (template: TemplateObject) => {
  return beautify(ssr(template), { format: 'html' })
}

test('ssr: Self closing - html', () => {
  expect(render(h('input', { type: 'text' }))).toMatchSnapshot()
})

test('ssr: Self closing - svg', () => {
  expect(render(h('polygon', { points: '0,100 50,25 50,75 100,0' }))).toMatchSnapshot()
})

test('ssr: Falsey - undefined', () => {
  expect(render(h('div', { children: undefined }))).toMatchSnapshot()
})

test('ssr: Falsey - null', () => {
  //@ts-expect-error: children is null
  expect(render(h('div', { children: null }))).toMatchSnapshot()
})

test('ssr: Falsey - false', () => {
  // @ts-expect-error: testing false
  expect(render(h('div', { children: false }))).toMatchSnapshot()
})

test('ssr: Not really Falsey - ""', () => {
  expect(render(h('div', { children: '' }))).toMatchSnapshot()
})

test('ssr: Not really Falsey - 0', () => {
  // @ts-expect-error: testing 0
  expect(render(h('div', { children: 0 }))).toMatchSnapshot()
})

test('ssr: Not really Falsey - NaN', () => {
  // @ts-expect-error: testing NaN
  expect(render(h('div', { children: NaN }))).toMatchSnapshot()
})

test('ssr: Bad template - NaN', () => {
  // @ts-expect-error: testing bad object
  expect(render(h('div', { children: { string: 'string' } }))).toMatchSnapshot()
})

test('ssr: Conditional', () => {
  expect(render(h('div', { children: true && 'hello' }))).toMatchSnapshot()
})

test('ssr: Style attribute', () =>
  expect(
    render(
      h('div', {
        style: { backgroundColor: 'blue', margin: '12px', '--cssVar': 'red' },
        children: 'styles',
      }),
    ),
  ).toMatchSnapshot())

test('ssr: p-trigger attribute', () =>
  expect(
    render(
      h('div', {
        'p-trigger': {
          click: 'random',
          focus: 'thing',
        },
        children: 'triggers',
      }),
    ),
  ).toMatchSnapshot())

test('ssr: Array of templates', () =>
  expect(
    render(
      h('ul', {
        children: Array.from(Array(10).keys()).map((n) => h('li', { children: `${n}` })),
      }),
    ),
  ).toMatchSnapshot())

test('ssr: Should throw with attribute starting with on', () => {
  expect(() => {
    h('div', {
      children: h('template', {
        shadowrootmode: 'closed',
        children: h('img', {
          src: 'nonexistent.png',
          onerror: "alert('xss!')",
        }),
      }),
    })
  }).toThrow()
})

test('ssr: should throw on script tag', () => {
  expect(() => {
    h('script', { type: 'module', src: 'main.js' })
  }).toThrow()
})

test('ssr: Should not throw on script tag with trusted attribute', () => {
  expect(render(h('script', { type: 'module', src: 'main.js', trusted: true }))).toMatchSnapshot()
})

test('ssr: Escapes children', () => {
  const scriptContent = `<script type="text/javascript">
const hostRegex = /^https?://([^/]+)/.*$/i;
const host = document.URL.replace(hostRegex, '$1');
const socket = new WebSocket(/);
const reload = () =>{
  location.reload();
  console.log('...reloading');
};
socket.addEventListener('message', reload);
console.log('[plaited] listening for file changes');
</script>`

  expect(render(h('div', { children: scriptContent }))).toMatchSnapshot()
})

test('ssr: Does not escape children when trusted', () => {
  const scriptContent = `<script type="text/javascript">
const hostRegex = /^https?://([^/]+)/.*$/i;
const host = document.URL.replace(hostRegex, '$1');
const socket = new WebSocket(/);
const reload = () =>{
  location.reload();
  console.log('...reloading');
};
socket.addEventListener('message', reload);
console.log('[plaited] listening for file changes');
</script>`

  expect(render(h('div', { trusted: true, children: scriptContent }))).toMatchSnapshot()
})

const Template: FunctionTemplate = (attrs) => h('template', attrs)

test('ssr: Non declarative shadow DOM template', () => {
  const List: FunctionTemplate = ({ children }) =>
    h('ul', {
      children: [
        Template({ children: h('span', { children: 'I am a span!!!' }) }),
        //@ts-expect-error: children is not undefined
        ...(Array.isArray(children) ? children : [children]),
      ],
    })

  expect(
    render(
      h(List, {
        children: Array.from(Array(10).keys()).map((n) => h('li', { children: `item-${n}` })),
      }),
    ),
  ).toMatchSnapshot()
})

test('Fragment', () => {
  expect(
    render(
      Fragment({
        children: Array.from(Array(6).keys())
          .reverse()
          .map((n) => h('li', { children: n > 0 ? `In ${n}` : 'Blast Off!!!' })),
      }),
    ),
  ).toMatchSnapshot()
})

const styles = css.create({
  nestedLabel: {
    fontWeight: 'bold',
  },
  nestedComponent: {
    display: 'flex',
    flexDirection: 'column',
  },
  slottedParagraph: {
    color: 'rebeccapurple',
  },
  topComponent: {
    display: 'block',
  },
  image: {
    width: '100%',
    aspectRatio: '16 / 9',
  },
})

const NestedCustomElement: FunctionTemplate = ({ children, ...props }) =>
  h('nested-component', {
    ...props,
    children: [
      Template({
        shadowrootmode: 'open',
        shadowrootdelegatesfocus: true,
        children: [
          h('span', {
            ...styles.nestedLabel,
            children: 'inside nested template',
          }),
          h('slot', { name: 'nested' }),
        ],
      }),
      //@ts-expect-error: children is not undefined
      ...(Array.isArray(children) ? children : [children]),
    ],
  })

test('createTemplate: CustomElement hoisting its styles', () => {
  const el = h(NestedCustomElement, {})
  expect({ content: render(el), stylesheets: el.stylesheets }).toMatchSnapshot()
})

test('createTemplate: CustomElement with declarative shadow dom & host styles', () => {
  const el = h(NestedCustomElement, { ...styles.nestedComponent })
  expect({ content: render(el), stylesheets: el.stylesheets }).toMatchSnapshot()
})

test('createTemplate: CustomElement with styled slotted component', () => {
  const el = h(NestedCustomElement, {
    ...styles.nestedComponent,
    children: [
      h('p', {
        slot: 'nested',
        ...styles.slottedParagraph,
        children: 'slotted paragraph',
      }),
    ],
  })
  expect({ content: render(el), stylesheets: el.stylesheets }).toMatchSnapshot()
})

const TopCustomElement: FunctionTemplate = ({ children, ...props }) =>
  h('top-component', {
    ...props,
    children: [
      Template({
        shadowrootdelegatesfocus: true,
        shadowrootmode: 'open',
        children: h(NestedCustomElement, {
          children: h('p', {
            slot: 'nested',
            ...styles.slottedParagraph,
            children: 'slotted paragraph',
          }),
        }),
      }),
      //@ts-expect-error: children is not undefined
      ...(Array.isArray(children) ? children : [children]),
    ],
  })

test('createTemplate: CustomElement with declarative shadow dom and nested declarative shadow dom', () => {
  const el = h(TopCustomElement, {})
  expect({ content: render(el), stylesheets: el.stylesheets }).toMatchSnapshot()
})

test('createTemplate: CustomElement with declarative shadow dom and nested declarative shadow dom plus host styles', () => {
  const el = h(TopCustomElement, { ...styles.topComponent })
  expect({ content: render(el), stylesheets: el.stylesheets }).toMatchSnapshot()
})

test('createTemplate: CustomElement with declarative shadow dom and nested declarative shadow dom plus host styles and child', () => {
  const el = h(TopCustomElement, {
    ...styles.topComponent,
    children: h('img', { ...styles.image }),
  })
  expect({ content: render(el), stylesheets: el.stylesheets }).toMatchSnapshot()
})

const hoistStyles = css.create({
  var1: {
    width: '100%',
  },
  var2: {
    width: '100%',
  },
  var3: {
    color: 'blue',
  },
})

test('ssr: Properly hoist and deduplicates multiple stylesheets on a single node', () => {
  expect(
    h('div', {
      ...css.assign(hoistStyles.var1, hoistStyles.var2, hoistStyles.var3),
    }).stylesheets.length,
  ).toBe(2)
})

test('ssr: filters out falsey style object', () => {
  expect(
    h('div', {
      ...css.assign(hoistStyles.var1, hoistStyles.var2, false, undefined, null),
    }).stylesheets.length,
  ).toBe(1)
})

test('createTemplate: Trims whitespace', () => {
  expect(
    render(
      h('div', {
        children: '   trims white-space    ',
      }),
    ),
  ).toMatchSnapshot()
})
