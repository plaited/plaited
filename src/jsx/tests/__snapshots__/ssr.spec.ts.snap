// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`useSSR: with imports and body 1`] = `
"<body class="body_18a0lbx p_mygp1 pv4i7qk p_wsdntf">
    <style>
        .p_mygp1{width:100%;}.pv4i7qk{padding:0;}.p_wsdntf{height:100vh;}
    </style>
    <div type="text"></div>
</body>"
`;

exports[`useSSR: with imports and body + head 1`] = `
"<head>
    <style>
        .p_mygp1{width:100%;}.pv4i7qk{padding:0;}.p_wsdntf{height:100vh;}
    </style>
</head>
<body class="body_18a0lbx p_mygp1 pv4i7qk p_wsdntf">
    <div type="text"></div>
</body>"
`;

exports[`ssr: Self closing - html 1`] = `"<input type="text" />"`;

exports[`ssr: Self closing - svg 1`] = `"<polygon points="0,100 50,25 50,75 100,0" />"`;

exports[`ssr: Falsey - undefined 1`] = `"<div></div>"`;

exports[`ssr: Falsey - null 1`] = `"<div></div>"`;

exports[`ssr: Falsey - false 1`] = `"<div></div>"`;

exports[`ssr: Not really Falsey - "" 1`] = `"<div></div>"`;

exports[`ssr: Not really Falsey - 0 1`] = `"<div>0</div>"`;

exports[`ssr: Not really Falsey - NaN 1`] = `"<div>NaN</div>"`;

exports[`ssr: Bad template - NaN 1`] = `"<div></div>"`;

exports[`ssr: Conditional 1`] = `"<div>hello</div>"`;

exports[`ssr: Style attribute 1`] = `"<div style="background-color:blue; margin:12px; --cssVar:red;">styles</div>"`;

exports[`ssr: p-trigger attribute 1`] = `"<div p-trigger="click:random focus:thing">triggers</div>"`;

exports[`ssr: Array of templates 1`] = `
"<ul>
    <li>0</li>
    <li>1</li>
    <li>2</li>
    <li>3</li>
    <li>4</li>
    <li>5</li>
    <li>6</li>
    <li>7</li>
    <li>8</li>
    <li>9</li>
</ul>"
`;

exports[`ssr: Should not throw on script tag with trusted attribute 1`] = `"<script type="module" " src="main.js " "></script>"`;

exports[`ssr: Escapes children 1`] = `
"<div>&lt;script type=&quot;text/javascript&quot;&gt; const hostRegex = /^https?://([^/]+)/.*$/i;
    const host = document.URL.replace(hostRegex, &#39;$1&#39;); const socket
    = new WebSocket(/); const reload = () =&gt;{ location.reload(); console.log(&#39;...reloading&#39;);
    }; socket.addEventListener(&#39;message&#39;, reload); console.log(&#39;[plaited]
    listening for file changes&#39;); &lt;/script&gt;</div>"
`;

exports[`ssr: Does not escape children when trusted 1`] = `
"<div>
    <script type="text/javascript">
        const hostRegex = /^https?://([^/]+)/.*$/i;
        const host = document.URL.replace(hostRegex, '$1');
        const socket = new WebSocket(/);
        const reload = () =>{
          location.reload();
          console.log('...reloading');
        };
        socket.addEventListener('message', reload);
        console.log('[plaited] listening for file changes');
    </script>
</div>"
`;

exports[`ssr: Non declarative shadow DOM template 1`] = `
"<ul>
    <template><span>I am a span!!!</span>
    </template>
    <li>item-0</li>
    <li>item-1</li>
    <li>item-2</li>
    <li>item-3</li>
    <li>item-4</li>
    <li>item-5</li>
    <li>item-6</li>
    <li>item-7</li>
    <li>item-8</li>
    <li>item-9</li>
</ul>"
`;

exports[`Fragment 1`] = `
"<li>In 5</li>
<li>In 4</li>
<li>In 3</li>
<li>In 2</li>
<li>In 1</li>
<li>Blast Off!!!</li>"
`;

exports[`createTemplate: CustomElement hoisting its styles 1`] = `
{
  "content": 
"<nested-component>
    <template shadowrootmode="open" shadowrootdelegatesfocus>
        <style>
            .p_ju9u1n{font-weight:bold;}
        </style><span class="nestedLabel__15w1ydc p_ju9u1n">inside nested template</span>
        <slot
        name="nested"></slot>
    </template>
</nested-component>"
,
  "stylesheets": [],
}
`;

exports[`createTemplate: CustomElement with declarative shadow dom & host styles 1`] = `
{
  "content": 
"<style>
    .p1rlu8mi{display:flex;}.p_8mez4t{flex-direction:column;}
</style>
<nested-component class="nestedComponent_ey8y76 p1rlu8mi p_8mez4t">
    <template shadowrootmode="open" shadowrootdelegatesfocus>
        <style>
            .p_ju9u1n{font-weight:bold;}
        </style><span class="nestedLabel__15w1ydc p_ju9u1n">inside nested template</span>
        <slot
        name="nested"></slot>
    </template>
</nested-component>"
,
  "stylesheets": [
    ".p1rlu8mi{display:flex;}",
    ".p_8mez4t{flex-direction:column;}",
  ],
}
`;

exports[`createTemplate: CustomElement with styled slotted component 1`] = `
{
  "content": 
"<style>
    .p1rlu8mi{display:flex;}.p_8mez4t{flex-direction:column;}.p91m869{color:rebeccapurple;}
</style>
<nested-component class="nestedComponent_ey8y76 p1rlu8mi p_8mez4t">
    <template shadowrootmode="open" shadowrootdelegatesfocus>
        <style>
            .p_ju9u1n{font-weight:bold;}
        </style><span class="nestedLabel__15w1ydc p_ju9u1n">inside nested template</span>
        <slot
        name="nested"></slot>
    </template>
    <p class="slottedParagraph_g28n4j p91m869" slot="nested">slotted paragraph</p>
</nested-component>"
,
  "stylesheets": [
    ".p1rlu8mi{display:flex;}",
    ".p_8mez4t{flex-direction:column;}",
    ".p91m869{color:rebeccapurple;}",
  ],
}
`;

exports[`createTemplate: CustomElement with declarative shadow dom and nested declarative shadow dom 1`] = `
{
  "content": 
"<top-component>
    <template shadowrootdelegatesfocus shadowrootmode="open">
        <style>
            .p91m869{color:rebeccapurple;}
        </style>
        <nested-component>
            <template shadowrootmode="open" shadowrootdelegatesfocus>
                <style>
                    .p_ju9u1n{font-weight:bold;}
                </style><span class="nestedLabel__15w1ydc p_ju9u1n">inside nested template</span>
                <slot
                name="nested"></slot>
            </template>
            <p class="slottedParagraph_g28n4j p91m869" slot="nested">slotted paragraph</p>
        </nested-component>
    </template>
</top-component>"
,
  "stylesheets": [],
}
`;

exports[`createTemplate: CustomElement with declarative shadow dom and nested declarative shadow dom plus host styles 1`] = `
{
  "content": 
"<style>
    .p131mv1i{display:block;}
</style>
<top-component class="topComponent_46img7 p131mv1i">
    <template shadowrootdelegatesfocus shadowrootmode="open">
        <style>
            .p91m869{color:rebeccapurple;}
        </style>
        <nested-component>
            <template shadowrootmode="open" shadowrootdelegatesfocus>
                <style>
                    .p_ju9u1n{font-weight:bold;}
                </style><span class="nestedLabel__15w1ydc p_ju9u1n">inside nested template</span>
                <slot
                name="nested"></slot>
            </template>
            <p class="slottedParagraph_g28n4j p91m869" slot="nested">slotted paragraph</p>
        </nested-component>
    </template>
</top-component>"
,
  "stylesheets": [
    ".p131mv1i{display:block;}",
  ],
}
`;

exports[`createTemplate: CustomElement with declarative shadow dom and nested declarative shadow dom plus host styles and child 1`] = `
{
  "content": 
"<style>
    .p131mv1i{display:block;}.p_mygp1{width:100%;}.p_x8x9vx{aspect-ratio:16 / 9;}
</style>
<top-component class="topComponent_46img7 p131mv1i">
    <template shadowrootdelegatesfocus shadowrootmode="open">
        <style>
            .p91m869{color:rebeccapurple;}
        </style>
        <nested-component>
            <template shadowrootmode="open" shadowrootdelegatesfocus>
                <style>
                    .p_ju9u1n{font-weight:bold;}
                </style><span class="nestedLabel__15w1ydc p_ju9u1n">inside nested template</span>
                <slot
                name="nested"></slot>
            </template>
            <p class="slottedParagraph_g28n4j p91m869" slot="nested">slotted paragraph</p>
        </nested-component>
    </template>
    <img class="image__c0ru3k p_mygp1 p_x8x9vx" />
</top-component>"
,
  "stylesheets": [
    ".p131mv1i{display:block;}",
    ".p_mygp1{width:100%;}",
    ".p_x8x9vx{aspect-ratio:16 / 9;}",
  ],
}
`;

exports[`ssr: filters out falsey classNames 1`] = `"<div class="truthy"></div>"`;

exports[`createTemplate: Trims whitespace 1`] = `"<div>trims white-space</div>"`;
